WebDAV camera

jkeljo opened the pull request:
## Breaking Change:

None

## Description:
As suggested by @dgomes in #19758, here's a camera backed by image files on
a WebDAV share. The entity exposes a stream and individual images, iterating
through the images in sorted order by filename at a configurable rate.

**Related issue (if applicable):** N/A

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#9085

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: webdav
    name: file_server
    host: 'https://server.name'
    image_interval: '1:00'
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [X] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

jkeljo committed:
WebDAV camera

As suggested by @dgomes in #19758, here's a camera backed by image files on
a WebDAV share. The entity exposes a stream and individual images, iterating
through the images in sorted order by filename at a configurable rate.

jkeljo committed:
lint

rohankapoorcom commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,242 @@
+
+    name = config[CONF_NAME]
+    host = config[CONF_HOST]
+    path = config.get(CONF_PATH)
```
Use [] for accessing required or default keys.

jkeljo replied:
Will fix.

rohankapoorcom commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,242 @@
+            [WebDavCamera(name, client, session, config[CONF_IMAGE_INTERVAL])],
+            update_before_add=True)
+    except WebDavException as exception:
+        _LOGGER.warning(
```
warning or error?

jkeljo replied:
[The docs](https://developers.home-assistant.io/docs/en/integration_quality_scale_index.html) specify to warn once when the device is unavailable, so I took that to mean warning.

rohankapoorcom replied:
Yep, that's my bad. Do you need the exception logged here? Since we know already what caused it and are logging a message already.

rohankapoorcom commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,242 @@
+                _LOGGER.info("Reconnected to WebDAV camera %s", self.name)
+        except WebDavException as exception:
+            self._files = []
+            if self._available:
```
I don't think this if check is needed before setting available to false.

jkeljo replied:
Same as above, preventing multiple logs.

rohankapoorcom commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,242 @@
+                           self._client.list(self._directory) if
+                           not filename.endswith('/')]
+            self._files.sort()
+            if not self._available:
```
I don't think this if check is needed before setting available to true.

jkeljo replied:
The guard is more for the log than the setting of `_available`, since [the docs](https://developers.home-assistant.io/docs/en/integration_quality_scale_index.html) say to log once on reconnect.

jkeljo committed:
Feedback

dgomes reviewed (commented):
Maybe instead of the wait_and_advance() you could use `async_track_time_change` from homeassistant.helpers.event

awarecan commented:
It is harsh time for new integration as we are doing refactoring, we are now requesting all components has to have a [manifest file](https://github.com/home-assistant/architecture/issues/179).

Please rebase you branch against current `dev` branch, and create a `manifest.json` file under your component's folder.

You may need to do more rebase in the next few days as we are still in the progress of refactoring.

jkeljo committed:
Use async_track_time_interval to handle advance intervals for me

jkeljo commented:
@dgomes, thanks for the pointer! I see `async_track_time_interval` in that same file, which seems to be more what I want.

@awarecan, no worries, I know how it goes. I can roll with it. :-)

jkeljo committed:
Merge remote-tracking branch 'upstream/dev' into webdav

jkeljo committed:
Upgrade to webdavclient3 0.11

jkeljo committed:
Add manifest file

dgomes commented:
Can you please rebase ? There several changes in the core that must be tested with this component before any merge.

jkeljo committed:
Merge remote-tracking branch 'upstream/dev' into webdav

Quentame commented on homeassistant/components/webdav/manifest.json:
```diff
@@ -0,0 +1,10 @@
+    "webdavclient3==0.11"
+  ],
+  "dependencies": [],
+  "codeowners": []
```
You should add yourself to the code owners :wink:

balloob commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,234 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+REQUIREMENTS = ['webdavclient3==0.11']
```
No need for this, it's part of the manifest.

balloob commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,234 @@
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up a single webdav camera."""
+    from webdav3.client import Client
```
imports can be done at the top of the file now.

balloob commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,234 @@
+        with await self._image_lock:
+            if self._image is not None:
+                return self._image
+            from aiohttp import ClientError
```
can import this at teh top.

balloob commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,234 @@
+    @property
+    def supported_features(self):
+        """Return supported features."""
+        return [SUPPORT_ON_OFF, SUPPORT_STREAM]
```
SUPPORT STREAM means you support RTSP. The built in API for streaming is not the same. Please remove this support flag and the `stream_source` property.

balloob commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,234 @@
+CONF_IMAGE_INTERVAL = "image_interval"
+CONF_KEY_PATH = "ssl_client_key"
+
+IMAGE_URL = '/api/camera_proxy/{0}?token={1}&frame={2}'
```
remove this.

MartinHjelmare commented:
@jkeljo are you planning to continue here?

jkeljo commented:
Yep, definitely. Just a matter of finding some free time, hopefully in the next couple weeks. If it’s cluttering things up for you, feel free to close it out and I’ll open another when I can get back to it.  



On July 14, 2019 at 9:04:32 AM, Martin Hjelmare (notifications@github.com(mailto:notifications@github.com)) wrote:

>  
> @jkeljo(https://github.com/jkeljo) are you planning to continue here?
>  
>  
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub(https://github.com/home-assistant/home-assistant/pull/22607?email_source=notifications&email_token=ABTZNWJL2BJCZED3HW37GJ3P7NFBBA5CNFSM4HCTBN3KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZ4IJNQ#issuecomment-511214774), or mute the thread(https://github.com/notifications/unsubscribe-auth/ABTZNWKKNISVGJS7RDEZ3LTP7NFBBANCNFSM4HCTBN3A).
>  
>  
>

MartinHjelmare commented:
@jkeljo please close, since then you can open this PR again yourself, when you have time, and we can keep the comment history in one place.

balloob commented:
oops, waiting for @jkeljo to close, so he can also re-open it, compared to when I close it.

jkeljo committed:
Merge remote-tracking branch 'upstream/dev' into webdav

jkeljo committed:
Feedback

jkeljo committed:
Regenerate CODEOWNERS

jkeljo committed:
Fix background crash in sisyphus integration

I think I was thinking in another language, since this was not valid Python. Result was that the thread that maintained the SocketIO connection to the table would die early on, so no status updates were actually flowing from the table.

jkeljo committed:
Update requirements

MartinHjelmare commented on homeassistant/components/sisyphus/__init__.py:
```diff
@@ -35,7 +35,7 @@ class SocketIONoiseFilter(logging.Filter):
 
         def filter(self, record):
-            if record.msg.contains("waiting for connection"):
+            if "waiting for connection" in record.msg:
```
This change should not be included in this PR. Please fix this in another PR.

jkeljo replied:
Snuck in there from being based on another PR that's already merged. Will clean up.

jkeljo committed:
Merge remote-tracking branch 'upstream/dev' into webdav

jkeljo committed:
Merge remote-tracking branch 'upstream/dev' into webdav

MartinHjelmare commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,235 @@
+This component models a WebDAV share full of images as a camera.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/camera.webdav/
```
We don't add the url to the docstring anymore. Just keep the very first line.

MartinHjelmare commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,235 @@
+"""
+
+import logging
+
```
Please remove this blank line. I recommend the isort tool to automatically sort and group imports.

MartinHjelmare commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,235 @@
+
+import asyncio
+from datetime import timedelta
+from aiohttp import ClientError
```
Please add a blank line between standard library and 3rd party import groups.

MartinHjelmare commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,235 @@
+            "webdav_root": path,
+        }
+    )
+    session = async_get_clientsession(hass)
```
This should be called in async context. Currently we're in sync context.

MartinHjelmare commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,235 @@
+SCAN_INTERVAL = timedelta(minutes=15)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
```
Change `def setup_platform` to `async def async_setup_platform`.

MartinHjelmare replied:
Please rename `add_entities` to `async_add_entities`.

MartinHjelmare commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,235 @@
+    session = async_get_clientsession(hass)
+
+    try:
+        client.list("/")  # This will throw if we can't access the share
```
Schedule this on the executor thread pool with `hass.async_add_executor_job`.

MartinHjelmare commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,235 @@
+            update_before_add=True,
+        )
+    except WebDavException as exception:
+        _LOGGER.warning("Failed to connect to %s: %s", client.get_url(""), exception)
```
Does `client.get_url` make I/O?

MartinHjelmare commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,235 @@
+        while images_checked < len(self._files):
+            self._image_number += 1
+            images_checked += 1
+            content_type = self._client.get_property(
```
This makes I/O so needs to be scheduled on the executor thread pool. Use `await self.hass.async_add_executor_job(...)`.

MartinHjelmare commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,235 @@
+        """Return True if the camera is playing through files in the share."""
+        return self._stop_advancing is not None
+
+    def turn_on(self):
```
This should be a coroutine and rename it to `async_turn_on`.

MartinHjelmare commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,235 @@
+                self.hass, self._advance, self._image_interval
+            )
+
+    def turn_off(self):
```
As above.

MartinHjelmare commented on homeassistant/components/webdav/camera.py:
```diff
@@ -0,0 +1,235 @@
+    async def async_added_to_hass(self):
+        """Set up periodic image advancement."""
+        self._image_lock = asyncio.Lock()
+        self.turn_on()
```
```py
await self.async_turn_on()
```

MartinHjelmare reviewed (changes requested):
Be wary of mixing sync and async calls in the same context. Coroutines should be awaited in async context and blocking (sync) I/O is not allowed in async context.

balloob commented:
This PR seems to have gone stale. Closing it.
